Generated Go code removes spaces from regexes

The `wbnf gen` command to generate Go code removes spaces from regexes that have been explicitly added with `\_`. All whitespace is stripped from regexes in wbnf to make regexes easier to understand by splitting up their parts. Most whitespace can be represented by backslash sequences, except space, for which wbnf has `\_` as a special extension.

However after `\_` has been replaced with a space after previously stripping whitespace, the code generator then strips all the spaces in `cmd/codegen/grammar.go:safeString`.

This results in a panic in the sysl grammar:
```
panic: regexp: Compile(`(?m)\A[]*(?:([^(\r?\n)]+))[]*`): error parsing regexp: unexpected ): `(?m)\A[]*(?:([^(\r?\n)]+))[]*`
```

The `[]` should have a space, and because the `]` appears after `[`, it does not close the character class, but instead puts `]` into the character class. This causes the mismatched parens because the character class does not close when it should.

The main function of `safeString` is to make the string a Go string that is syntactically valid. This can be done with `fmt.Sprintf("%#v", s)` which is likely safer against any mistakes (and clearer to Go programmers). If the ast has whitespace in it, we should leave it there. I can see no reason why the go generate should remove it.

--- a/cmd/codegen/grammar.go
+++ b/cmd/codegen/grammar.go
@@ -45,7 +45,7 @@ func (g *goNode) Add(n goNode) {
 }
 
 func safeString(src string) string {
-	r := strings.NewReplacer("`", "`+\"`\"+`", " ", "", "\n", "")
+	r := strings.NewReplacer("`", "`+\"`\"+`")
 	return r.Replace(src)
 }
 

--- a/cmd/codegen/grammar_test.go
+++ b/cmd/codegen/grammar_test.go
@@ -1,0 +1,30 @@
+package codegen
+
+import (
+	"testing"
+
+	"github.com/arr-ai/wbnf/parser"
+	"github.com/stretchr/testify/assert"
+)
+
+// TestWalkTermTerminal tests the the output of terminal nodes in AST (those
+// with no children (noScope).
+func TestWalkTermTerminal(t *testing.T) {
+	// Note: These tests are far from comprehensive. They include enough
+	// to test fixes for #59 (Generated Go code removes spaces from regexes)
+	tests := map[string]struct {
+		input  parser.Term
+		output string
+	}{
+		"S simple":  {parser.S("hello"), "parser.S(`hello`)"},
+		"S newline": {parser.S("\n"), "parser.S(`\n`)"},
+		"RE simple": {parser.RE("[A-Z]+"), "parser.RE(`[A-Z]+`)"},
+		"RE space":  {parser.RE("[ ]+"), "parser.RE(`[ ]+`)"},
+	}
+	for name, tt := range tests {
+		tt := tt
+		t.Run(name, func(t *testing.T) {
+			assert.Equal(t, tt.output, walkTerm(tt.input).name)
+		})
+	}
+}

